pkg/virt-launcher/virtwrap/device/sriov/hostdev.go

```diff
@@ -85,3 +92,102 @@ func createHostDevice(iface v1.Interface, hostPCIAddress string) (*api.HostDevic
 
 	return hostDev, nil
 }
+
+type deviceDetacher interface {
+	DetachDevice(xmlData string) error
+}
+
+type eventRegistrar interface {
+	Register() error
+	Deregister() error
+	EventChannel() <-chan interface{}
+}
+
+func SafelyDetachHostDevices(domainSpec *api.DomainSpec, eventDetach eventRegistrar, dom deviceDetacher, timeout time.Duration) error {
+	sriovDevices := FilterHostDevices(domainSpec)
+	if len(sriovDevices) == 0 {
+		log.Log.Info("No SR-IOV host-devices to detach.")
+		return nil
+	}
+
+	if err := eventDetach.Register(); err != nil {
+		return fmt.Errorf("failed to detach host-devices: %v", err)
+	}
+	defer func() {
+		if err := eventDetach.Deregister(); err != nil {
+			log.Log.Reason(err).Errorf("failed to detach host-devices: %v", err)
+		}
+	}()
+
+	if err := detachHostDevices(dom, sriovDevices); err != nil {
+		return err
+	}
+
+	return waitHostDevicesToDetach(eventDetach, sriovDevices, timeout)
+}
+
+func FilterHostDevices(domainSpec *api.DomainSpec) []api.HostDevice {
+	var hostDevices []api.HostDevice
+
+	for _, hostDevice := range domainSpec.Devices.HostDevices {
+		if hostDevice.Alias != nil && strings.HasPrefix(hostDevice.Alias.GetName(), AliasPrefix) {
+			hostDevices = append(hostDevices, hostDevice)
+		}
+	}
+	return hostDevices
+}
+
+func detachHostDevices(dom deviceDetacher, hostDevices []api.HostDevice) error {
+	for _, hostDev := range hostDevices {
+		devXML, err := xml.Marshal(hostDev)
+		if err != nil {
+			return fmt.Errorf("failed to encode (xml) hostdev %v, err: %v", hostDev, err)
+		}
+		err = dom.DetachDevice(string(devXML))
+		if err != nil {
+			return fmt.Errorf("failed to detach hostdev %s, err: %v", devXML, err)
+		}
+		log.Log.Infof("Successfully hot-unplug hostdev: %s (%v)", hostDev.Alias.GetName(), hostDev.Source.Address)
```

Comment:
Could you please change "hostdev" to "host-devices" as well?

Reply:
I prefer to use at this low level the name used in the xml.
I answered about this in a previous [comment](https://github.com/kubevirt/kubevirt/pull/4741#discussion_r578820319).